c10/cuda/CUDAStream.cpp

```diff
@@ -304,7 +306,47 @@ LeakyStreamInternals* CUDAStream_internals(CUDAStream s) {
   }
 }
 
+void check_assert_state(const LeakyStreamInternals* ptr) {
+  if (!ptr->assert_state) {
+    return;
+  }
+
+  if (ptr->assert_state->error != 0) {
+    CUDAGuard device_guard(ptr->device_index);
+
+    // wait for kernel to complete
+    C10_CUDA_CHECK(cudaDeviceSynchronize());
+    CUDAAssert assert_state(*ptr->assert_state);    // create local isolation copy
+
+    // reset assert state
+    std::memset(ptr->assert_state, 0, sizeof(CUDAAssert));
+
+    switch (assert_state.type) {
+      case CUDAAssertKind::ASSERTION_FAILED:
+        // assertion failed
+        throw ::c10::Error({"unknown", assert_state.file, assert_state.line}, assert_state.message);
+
+      case CUDAAssertKind::INDEX_OUT_OF_BOUNDS: {
+        // index out of bounds
+        auto index_error = reinterpret_cast<CUDAAssertDetailIndexKernel*>(assert_state.details);
+        throw ::c10::IndexError({"unknown", assert_state.file, assert_state.line},
+          ::c10::str("index ", index_error->index, " is out of bounds for dimension ", index_error->axis, " with size ", index_error->size));
+        break;
```

Comment:
The break is not needed here, but if you have one here, add them to the other cases as well.

Reply:
> I wonder, `check_assert_state` is only called from `CUDAStream_fromInternals`. When is that called?

@pietern `from_Internals` is called for example for every call to `getDefaultCUDAStream()`. This is a quite central place for the check but will be executed mostly once during setup of a kernel launch. I did not expose the `check_assert_state` function in the prototype, but it could simply become a member of `CUDAStream` to be called explicitly, e.g. by C++ clients.

I must admit that I do not yet fully understand the architecture of c10, e.g. I had no contact to `HIP` etc. In this context it is for example unclear to me whether Stream sub-classes are allowed to expose new functionality like cuda-specific assert/error handling.